File: server/utils/locationMatcher.js
import User from '../models/User.js';

/**
 * Calculate distance between two coordinates using Haversine formula
 * @param {number} lat1 - Latitude of first point
 * @param {number} lon1 - Longitude of first point
 * @param {number} lat2 - Latitude of second point
 * @param {number} lon2 - Longitude of second point
 * @returns {number} Distance in kilometers
 */
export const calculateDistance = (lat1, lon1, lat2, lon2) => {
  const R = 6371; // Radius of the Earth in kilometers
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = 
    Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * 
    Math.sin(dLon/2) * Math.sin(dLon/2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
  return R * c;
};

/**
 * Check if two locations match (same city and state, or within specified radius)
 * @param {Object} location1 - First location object
 * @param {Object} location2 - Second location object
 * @param {number} maxDistance - Maximum distance in km for matching (default: 50)
 * @returns {boolean} True if locations match
 */
export const locationsMatch = (location1, location2, maxDistance = 50) => {
  // Exact city and state match
  if (location1.city?.toLowerCase() === location2.city?.toLowerCase() &&
      location1.state?.toLowerCase() === location2.state?.toLowerCase()) {
    return true;
  }

  // Distance-based matching if coordinates are available
  if (location1.coordinates && location2.coordinates) {
    const distance = calculateDistance(
      location1.coordinates.latitude,
      location1.coordinates.longitude,
      location2.coordinates.latitude,
      location2.coordinates.longitude
    );
    return distance <= maxDistance;
  }

  // Fallback: check if cities are in the same state and nearby
  if (location1.state?.toLowerCase() === location2.state?.toLowerCase()) {
    // For same state, we'll consider it a match if cities are similar
    // This is a simplified approach - in production you might want more sophisticated matching
    return true;
  }

  return false;
};

/**
 * Find MSMEs that match the consignment's origin location
 * @param {Object} consignmentOrigin - Origin location of the consignment
 * @param {number} maxDistance - Maximum distance in km for matching
 * @returns {Array} Array of matching MSME users
 */
export const findMatchingMSMEs = async (consignmentOrigin, maxDistance = 50) => {
  try {
    // Get all MSME users
    const msmeUsers = await User.find({ companyType: 'msme' });
    
    const matchingMSMEs = msmeUsers.filter(user => {
      if (!user.location) return false;
      
      return locationsMatch(consignmentOrigin, user.location, maxDistance);
    });

    return matchingMSMEs;
  } catch (error) {
    console.error('Error finding matching MSMEs:', error);
    return [];
  }
};

/**
 * Get consignments that match an MSME's location
 * @param {Object} msmeLocation - MSME's location
 * @param {Array} consignments - Array of consignments to filter
 * @param {number} maxDistance - Maximum distance in km for matching
 * @returns {Array} Array of matching consignments
 */
export const getMatchingConsignments = (msmeLocation, consignments, maxDistance = 50) => {
  if (!msmeLocation) return consignments; // If no location, show all

  return consignments.filter(consignment => {
    return locationsMatch(consignment.origin, msmeLocation, maxDistance);
  });
};

/**
 * Normalize location string to extract city and state
 * @param {string} locationString - Location string (e.g., "Mumbai, Maharashtra")
 * @returns {Object} Normalized location object
 */
export const normalizeLocation = (locationString) => {
  if (!locationString) return null;

  const parts = locationString.split(',').map(part => part.trim());
  
  if (parts.length >= 2) {
    return {
      city: parts[0],
      state: parts[1],
      fullAddress: locationString
    };
  } else if (parts.length === 1) {
    return {
      city: parts[0],
      state: parts[0], // Fallback to same as city
      fullAddress: locationString
    };
  }

  return null;
};

/**
 * Get location-based recommendations for MSMEs
 * @param {string} msmeId - MSME user ID
 * @returns {Object} Recommendations object
 */
export const getLocationRecommendations = async (msmeId) => {
  try {
    const msme = await User.findById(msmeId);
    if (!msme || msme.companyType !== 'msme' || !msme.location) {
      return { matchingConsignments: [], nearbyMSMEs: [] };
    }

    // Find nearby MSMEs (for potential partnerships)
    const nearbyMSMEs = await User.find({
      companyType: 'msme',
      _id: { $ne: msmeId },
      location: { $exists: true }
    });

    const nearbyPartners = nearbyMSMEs.filter(user => 
      locationsMatch(msme.location, user.location, 100) // 100km radius for partnerships
    );

    return {
      matchingConsignments: [], // Will be populated by the calling function
      nearbyMSMEs: nearbyPartners.slice(0, 5) // Top 5 nearby partners
    };
  } catch (error) {
    console.error('Error getting location recommendations:', error);
    return { matchingConsignments: [], nearbyMSMEs: [] };
  }
}; 

File: python_backend/utils/location_matcher.py
import math
from typing import Dict, List, Optional, Union, Any


def calculate_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate distance between two coordinates using Haversine formula
    Args:
        lat1: Latitude of first point
        lon1: Longitude of first point
        lat2: Latitude of second point
        lon2: Longitude of second point
    Returns:
        Distance in kilometers
    """
    R = 6371  # Radius of the Earth in kilometers
    d_lat = (lat2 - lat1) * math.pi / 180
    d_lon = (lon2 - lon1) * math.pi / 180
    a = (
        math.sin(d_lat/2) * math.sin(d_lat/2) +
        math.cos(lat1 * math.pi / 180) * math.cos(lat2 * math.pi / 180) * 
        math.sin(d_lon/2) * math.sin(d_lon/2)
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    return R * c


def normalize_location(location_string: str) -> Optional[Dict[str, str]]:
    """
    Normalize location string to extract city and state
    Args:
        location_string: Location string (e.g., "Mumbai, Maharashtra")
    Returns:
        Normalized location object
    """
    if not location_string:
        return None

    parts = [part.strip() for part in location_string.split(',')]
    
    if len(parts) >= 2:
        return {
            "city": parts[0],
            "state": parts[1],
            "fullAddress": location_string
        }
    elif len(parts) == 1:
        return {
            "city": parts[0],
            "state": parts[0],  # Fallback to same as city
            "fullAddress": location_string
        }

    return None


def locations_match(location1: Union[Dict, str], location2: Union[Dict, str], max_distance: float = 50) -> bool:
    """
    Check if two locations match (same city and state, or within specified radius)
    Args:
        location1: First location object or string
        location2: Second location object or string
        max_distance: Maximum distance in km for matching (default: 50)
    Returns:
        True if locations match
    """
    # Handle string locations by normalizing them
    loc1 = normalize_location(location1) if isinstance(location1, str) else location1
    loc2 = normalize_location(location2) if isinstance(location2, str) else location2

    if not loc1 or not loc2:
        return False

    # Exact city and state match
    if (loc1.get("city", "").lower() == loc2.get("city", "").lower() and
        loc1.get("state", "").lower() == loc2.get("state", "").lower()):
        return True

    # Distance-based matching if coordinates are available
    if loc1.get("coordinates") and loc2.get("coordinates"):
        distance = calculate_distance(
            loc1["coordinates"]["latitude"],
            loc1["coordinates"]["longitude"],
            loc2["coordinates"]["latitude"],
            loc2["coordinates"]["longitude"]
        )
        return distance <= max_distance

    # Fallback: check if cities are in the same state and nearby
    if loc1.get("state", "").lower() == loc2.get("state", "").lower():
        # For same state, we'll consider it a match if cities are similar
        # This is a simplified approach - in production you might want more sophisticated matching
        return True

    return False


def find_matching_msmes(consignment_origin: Union[Dict, str], users: List[Dict], max_distance: float = 50) -> List[Dict]:
    """
    Find MSMEs that match the consignment's origin location
    Args:
        consignment_origin: Origin location of the consignment
        users: Array of users from in-memory storage
        max_distance: Maximum distance in km for matching
    Returns:
        Array of matching MSME users
    """
    try:
        # Get all MSME users from the provided users array
        msme_users = [user for user in users if user.get("userType") == "msme"]
        
        matching_msmes = [
            user for user in msme_users 
            if user.get("location") and locations_match(consignment_origin, user["location"], max_distance)
        ]

        return matching_msmes
    except Exception as error:
        print(f"Error finding matching MSMEs: {error}")
        return []


def get_matching_consignments(msme_location: Union[Dict, str], consignments: List[Dict], max_distance: float = 50) -> List[Dict]:
    """
    Get consignments that match an MSME's location
    Args:
        msme_location: MSME's location
        consignments: Array of consignments to filter
        max_distance: Maximum distance in km for matching
    Returns:
        Array of matching consignments
    """
    if not msme_location:
        return consignments  # If no location, show all

    return [
        consignment for consignment in consignments
        if locations_match(consignment.get("origin"), msme_location, max_distance)
    ]


def get_location_recommendations(msme_id: str, users: List[Dict]) -> Dict[str, List]:
    """
    Get location-based recommendations for MSMEs
    Args:
        msme_id: MSME user ID
        users: Array of users from in-memory storage
    Returns:
        Recommendations object
    """
    try:
        msme = next((user for user in users if user.get("id") == msme_id), None)
        if not msme or msme.get("userType") != "msme" or not msme.get("location"):
            return {"matchingConsignments": [], "nearbyMSMEs": []}

        # Find nearby MSMEs (for potential partnerships)
        nearby_msmes = [
            user for user in users
            if user.get("userType") == "msme" and 
            user.get("id") != msme_id and 
            user.get("location")
        ]

        nearby_partners = [
            user for user in nearby_msmes
            if locations_match(msme["location"], user["location"], 100)  # 100km radius for partnerships
        ]

        return {
            "matchingConsignments": [],  # Will be populated by the calling function
            "nearbyMSMEs": nearby_partners[:5]  # Top 5 nearby partners
        }
    except Exception as error:
        print(f"Error getting location recommendations: {error}")
        return {"matchingConsignments": [], "nearbyMSMEs": []} 